### PR TITLE
wpgu: Initial implementation of PixelBender shader execution

### PR DESCRIPTION
**Pull request by Aaron1011**
The implementation is split across four crates:
* `ruffle_render` now holds the main PixelBender bytecode parsing implementation (previously, this was in `ruffle_core`).
* `ruffle_core` holds some helper functions for converting between AVM2 `Value`s and the PixelBender vector types.
* `naga-pixelbender` (newly created) constructs a Naga `Module` from parsed PixelBender bytecode
* `ruffle_render_wgpu` sets up the render pipeline for the shader constructed by `naga-pixelbender`, and actually executes the shader.

The Actionscript-side shader parameters are passed in through uniforms. This allows us to cache the compiled `naga::Module` and associated wgpu types inside `ShaderData`, when it's first created. Each invocation of a `ShaderJob` only needs to create a bind group and render pass.

Limitations:

* Only a few of the PixelBender opcodes are implemented - however, this is enough to get Stemlands cannon rotation working, as well as a cool "donut" shader that I found and included as a test.
* PixelBender matrix types are not supported.
* Only BitmapData is supported as an input/output type - Flash Player also supports using Vector and ByteArray
* ShaderJob execution is always synchronous.

### CODE REVIEW

File: core/src/avm2/globals/flash/display/shader_job.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    let shader = this
+        .get_public_property("shader", activation)?
+        .as_object()
+        .expect("Missing Shader object");
+    let shader_data = shader
+        .get_public_property("data", activation)?
+        .as_object()
+        .expect("Missing ShaderData object")
+        .as_shader_data()
+        .expect("ShaderData object is not a ShaderData instance");
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/shader_job.rs:39]: Since these could be `null`, would it be better to do null handling rather than panic?

Comment (Aaron1011) [reply core/src/avm2/globals/flash/display/shader_job.rs:39]: This PR is already very large, so I'm postponing nicer error handling for a follow-up PR.

File: render/src/pixel_bender.rs [hunk 2]
```diff
@@ -0,0 +1,636 @@
+//! Pixel bender bytecode parsing code.
+//! This is heavling based on https://github.com/jamesward/pbjas and https://github.com/HaxeFoundation/format/tree/master/format/pbj
```
Comment (Lord-McSweeney) [inline render/src/pixel_bender.rs:2]: nit: Typo: "heavling" --> "heavily"

File: core/src/avm2/globals/flash/display/shader_job.rs (hunk 1 above)
Comment (Lord-McSweeney) [reply core/src/avm2/globals/flash/display/shader_job.rs:39]: Could you add some TODO notices in that case?

**Comment by Lord-McSweeney** [comment]
Is there any chance #11252 could occur with the naga compilation?

**Comment by Aaron1011** [comment]
@Lord-McSweeney Fortunately, no - PixelBender only allows you to write a fragment shader. The vertex shader is provided by us, and only has a single output (the fragment position).

File: core/src/avm2/globals/flash/display/ShaderJob.as [hunk 3]
```diff
@@ -34,12 +37,21 @@ package flash.display {
...
 
         public function get shader():Shader {
-            stub_getter("flash.display.ShaderJob", "shader");
-            return null;
+            return this._shader;
         }
 
         public function set shader(value:Shader):void {
+            this._shader = value;
             stub_setter("flash.display.ShaderJob", "shader");
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display/ShaderJob.as:45]: This isn't a stub anymore

File: core/src/avm2/globals/flash/display/ShaderJob.as [hunk 4]
```diff
@@ -34,12 +37,21 @@ package flash.display {
...
         }
+
+        public function get target():Object {
+            return this._target;
+        }
+
+        public function set target(value:Object):void {
+            this._target = value;
+            stub_setter("flash.display.ShaderJob", "target");
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display/ShaderJob.as:54]: Same here, if we're setting it, it's not a stub

File: desktop/src/main.rs [hunk 5]
```diff
@@ -67,6 +67,9 @@ fn init() {
     let subscriber = tracing_subscriber::fmt::Subscriber::builder()
         .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
         .finish();
+
+    env_logger::init();
```
Comment (Dinnerbone) [inline desktop/src/main.rs:71]: What's this for?

Comment (Aaron1011) [reply desktop/src/main.rs:71]: This gives us logs from naga, which uses the `log` crate

Comment (Dinnerbone) [reply desktop/src/main.rs:71]: We already should be seeing those, we redirect log to tracing. This would replace that redirection

Comment (Aaron1011) [reply desktop/src/main.rs:71]: It looks like we're only using `tracing_log` on web. I'll remove `env_logger` from this pr, and enable `tracing_log` in a separate PR.

Comment (Dinnerbone) [reply desktop/src/main.rs:71]: `tracing_subscriber` uses `tracing_log` by default, perhaps something's going wrong there? But I'm sure I've seen wgpu logs on desktop and those are log.

**Review by Dinnerbone — APPROVED** [review]

**Pull request merged**